signing_service/src/signing_service.py

```diff
@@ -1,60 +1,169 @@
+        If there was an unrecognized exception, it logs it and report to Sentry, then reraise and crash.
+        """
+
+        def _set_was_sigterm_caught_true(signum, frame):  # pylint: disable=unused-argument
+            logger.info('Closing connection and exiting on SIGTERM.')
+            self.was_sigterm_caught = True
+
+        # Handle shutdown signal.
+        signal.signal(signal.SIGTERM, _set_was_sigterm_caught_true)
+
+        while not self._was_sigterm_caught():
+            with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as tcp_socket:
+                try:
+                    self._connect(tcp_socket)
+                except socket.error as exception:
+                    logger.error(f'Socket error occurred: {exception}')
+
+                    # Only predefined list of exceptions should cause reconnect, others should be reraised.
+                    if isinstance(exception.args, tuple) and exception.args[0] not in SIGNING_SERVICE_RECOVERABLE_ERRORS:  # type: ignore
+                        raise
+
+                    # Increase delay and reconnect if the connection is interrupted due to a failure.
+                    self._increase_delay()
+                except KeyboardInterrupt:
+                    # Handle keyboard interrupt.
+                    logger.info('Closing connection and exiting on KeyboardInterrupt.')
+                    break
+                except Exception as exception:
+                    # If there was an unrecognized exception, log it and report to Sentry.
+                    crash_logger.error(f'Unrecognized exception occurred: {exception}')
+                    raise
+
+    def _connect(self, tcp_socket: socket.socket) -> None:
+        """ Creates socket and connects to given HOST and PORT. """
+        if self.current_reconnect_delay is not None:
+            logger.warning(f'Waiting {self.current_reconnect_delay} before connecting.')
+            sleep(self.current_reconnect_delay)
+
+        logger.warning(f'Connecting to {self.host}:{self.port}.')
+        tcp_socket.connect((self.host, self.port))
+        logger.warning(f'Connection established.')
 
-    socket_to_concent.send(b'GET / HTTP/1.0\r\n\r\n')
-    data = socket_to_concent.recv(1024)
-    logger.info(data)
-    socket_to_concent.close()
+        # Reset delay on successful connection and set flag that connection is established.
+        self.current_reconnect_delay = None
+        self._handle_connection(tcp_socket)
 
+    def _handle_connection(self, tcp_socket: socket.socket) -> None:  # pylint: disable=no-self-use
+        """ Inner loop that handles data exchange over socket. """
+        while True:
+            data = tcp_socket.recv(1024)
+            tcp_socket.send(data)
 
-def _parse_arguments():
+    def _increase_delay(self) -> None:
+        """ Increase current delay if connection cannot be established. """
+        assert self.current_reconnect_delay != 0
```

Comment:
Why not `> 0`?

Reply:
Because it can be also None but it can also be 0 so I have must not updated this line for some time. :+1: done

Reply:
It should not. But I mean that your assert was allowing negative values. They aren't valid, are they?

Reply:
As discussed, we allow 0.
Yes, it should not be negative but it can be None. 
Fixed already.